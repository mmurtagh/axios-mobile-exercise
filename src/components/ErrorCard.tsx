import React from 'react';
import styled from 'styled-components/native';

import { Button, Card, Paragraph } from '../components/components';
import { spacing } from '../utils/styling';

const ErrorButton = styled(Button)`
  margin-top: ${spacing()};  
`

const Container = styled(Card)`
  margin-top: ${spacing()};
`

const ErrorText = styled(Paragraph)`
  text-align: center;
`

/** 
 * @name: ErrorCard
 * @description: Component that informs the user there has been an error
 * and allows them to retry the action that caused the error
 * @param onPress: The function that is be called when the user presses the "Try Again" button
*/
export const ErrorCard = ({ onPress }: { onPress: () => any}) => {
  return (
    <Container testID="error-card">
      <ErrorText testID="error-text">There was an error with your request.</ErrorText>
      <ErrorButton testID="error-button" onPress={onPress} title="Try Again" />
    </Container>
  )
}
