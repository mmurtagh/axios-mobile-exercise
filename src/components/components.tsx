import styled from 'styled-components/native';
import MaterialIcon from 'react-native-vector-icons/MaterialIcons';
import React from 'react';
import { TouchableOpacityProps } from 'react-native';
import { IconProps } from 'react-native-vector-icons/Icon';
import { SafeAreaViewProps } from 'react-native-safe-area-context';

import {
  spacing,
  fontSize,
  fontColor,
  backdropColor,
  surfaceColor,
  primary,
} from '../utils/styling';

const SafeAreaView = styled.SafeAreaView`
  background-color: ${backdropColor};
  flex-grow: 1;
`;

const Container = styled.View`
  padding-horizontal: ${spacing()};
`;

const Touchable = styled.TouchableOpacity`
  background-color: ${primary};
  padding: ${spacing()}
  border-radius: 5px;
  flex-direction: row;
  align-items: center;
  justify-content: center;
`

interface AppButtonProps extends TouchableOpacityProps {
  title: string;
  icon?: string;
}

/** 
 * @name: Card
 * @description: Styled card component
*/
export const Card = styled.View`
  padding: ${spacing()};
  box-shadow: 0px 2px 1px rgba(0, 0, 0, 0.25);
  background-color: ${surfaceColor};
  border-radius: 5px;
`;

/** 
 * @name: Image
 * @description: Styled image component
*/
export const Image = styled.Image`
  aspect-ratio: ${({ aspectRatio }: { aspectRatio: number }) => aspectRatio};
  flex: 2;
  border-radius: 5px;
`;

/** 
 * @name: AppText
 * @description: Base component of all typography
*/
const AppText = styled.Text`
  font-family: Helvetica;
  color: ${fontColor};
`;

/** 
 * @name: HeadLine
 * @description: Large text
*/
export const Headline = styled(AppText)`
  font-size: ${fontSize('lg')};
`;

/** 
 * @name: Paragraph
 * @description: Medium text
*/
export const Paragraph = styled(AppText)`
  font-size: ${fontSize()};
`;

/** 
 * @name: Caption
 * @description: Small text
*/
export const Caption = styled(AppText)`
  font-size: ${fontSize('sm')};
`;

const ButtonText = styled(Paragraph)`
  padding-left: ${spacing('sm')};
  text-align: center;
  color: white;
`

/** 
 * @name: Screen
 * @description: Component that wraps all screens
*/
export const Screen = ({ children, ...props }: SafeAreaViewProps) => {
  return (
    <SafeAreaView {...props}>
      <Container>
        {children}
      </Container>
    </SafeAreaView>
  );
};

/** 
 * @name: Icon
 * @description: Vector icon from the MateriaIcons icon set
*/
export const Icon = (
  { name, size = 30, onPress, color = primary, ...props }: IconProps) => {
  return (
    <MaterialIcon
      {...props}
      color={color}
      size={size}
      name={name}
      onPress={onPress}
    />
  )
};

/** 
 * @name: Button
 * @description: Styled button
*/
export const Button = ({ title, onPress, icon, ...props }: AppButtonProps) => {
  return (
    <Touchable accessibilityRole="button" onPress={onPress} {...props}>
      {icon && <Icon size={20} color="white" name={icon} />}
      <ButtonText>{title}</ButtonText>
    </Touchable>
  )
}
