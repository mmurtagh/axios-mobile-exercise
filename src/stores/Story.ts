import { makeAutoObservable, toJS } from 'mobx';
import dayjs from 'dayjs';

export enum SupportedEntityType {
  LINK = 'LINK',
}

export enum SupportedBlockType {
  UNSTYLED = 'unstyled',
  UNORDERED_LIST_ITEM = 'unordered-list-item',
}

export enum SupportedInlineStyle {
  BOLD = 'BOLD',
}

/** 
 * @name sanitizeContentInstance
 * @description: Filters out unsupported features from raw AxiosContentInstances.
 * @param _contentInstance: The AxiosContentInstance object that needs to be sanitized.
*/
function sanitizeContentInstance (_contentInstance: AxiosContentInstance): AxiosContentInstance {
  const contentInstance = { ..._contentInstance }
  const { blocks: { blocks } } = contentInstance;

  // Filter out all blocks with unsupported block type
  contentInstance.blocks.blocks = blocks.filter(({ type }) => {
    const supportedTypes: string[] = Object.values(SupportedBlockType);

    return supportedTypes.includes(type)
  })

  return contentInstance
}

/** 
 * @class: Story
 * @description: An Axios story. A mobx domain object managed by StoryStore.js.
*/
export class Story {
  contentInstance: AxiosContentInstance;

  constructor(contentInstance: AxiosContentInstance) {
    this.contentInstance = sanitizeContentInstance(contentInstance);
    makeAutoObservable(this);
  }

  get headline(): string {
    return this.contentInstance.headline;
  }

  get publishedDate(): dayjs.Dayjs {
    return dayjs(this.contentInstance.published_date);
  }

  /** 
   * @name getImageSources
   * @description: Gets the list of image sources of a given ratio.
   * @param ratio: The desired image's aspect ratio
  */
   getImageSources (ratio: '1x1' | '4x3' | '16x9'): Crop[] {
    let image = this.contentInstance.primary_image

    if (image === null || !image.crops[ratio].sizes) {
      image = this.contentInstance.social_image;
    }

    if (image === null) {
      return []
    }

    return toJS(image.crops[ratio].sizes)
  }

  get id(): string {
    return this.contentInstance.id;
  }

  get author(): string | null {
    if (!this.contentInstance.authors.length) {
      return null;
    }

    return this.contentInstance.authors[0].display_name;
  }

  get primaryTopicName(): string | null {
    const topics = this.contentInstance.topics;

    if (topics === null || topics.length === 0) {
      return null;
    }

    return topics[0].name;
  }

  get blocks(): Block[] {
    return this.contentInstance.blocks.blocks;
  }

  get entityMap(): Entity[] {
    return this.contentInstance.blocks.entityMap;
  }

  /** 
   * @name imageDescription
   * @description: Gets a description of the image associated with the story.
  */
  get imageDescription(): string {
    let image = this.contentInstance.primary_image

    if (image === null) {
      image = this.contentInstance.social_image;
    }

    if (image === null) {
      return ''
    }

    const { blocks } = image.caption
    
    if (blocks.length === 0) {
      return ''
    }

    return blocks[0].text;
  }
}
