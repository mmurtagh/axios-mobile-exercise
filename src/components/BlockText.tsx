import React from "react";
import styled from "styled-components/native";
import { observer } from "mobx-react-lite";

import { Paragraph } from "./components";
import { spacing } from "../utils/styling";
import { SupportedBlockType } from '../stores/Story'

const BlockParagraph = styled(Paragraph)`
  margin-bottom: ${spacing()};
  flex: 1;
`

const Container = styled.View`
  flexDirection: row;
`

const Bullet = styled.View`
  width: 8px;
  height: 8px;
  border-radius: 4px;
  background-color: #444;
  margin-top: ${spacing('sm')};
  margin-right: ${spacing()};
`

/** 
 * @name: BlockText
 * @description: Renders a block of text either unstyled or as an unordered list item
 * @param block: The Block object representing the block of text
*/
export const BlockText = observer(({ block }: { block: Block }) => {
  return (
    <Container>
      {block.type === SupportedBlockType.UNORDERED_LIST_ITEM && <Bullet testID="bullet"/>}
      <BlockParagraph testID="block-paragraph">{block.text}</BlockParagraph>
    </Container>
  )
});
