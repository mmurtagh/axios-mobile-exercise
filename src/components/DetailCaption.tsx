import React from 'react'
import styled from 'styled-components/native'
import { observer } from 'mobx-react-lite';

import { Story } from '../stores/Story';
import { Caption } from './components';
import { spacing } from '../utils/styling';

const StyledCaption = styled(Caption)`
  padding-top: ${spacing()};
`

/** 
 * @name: DetailCaption
 * @description: Text displaying a stores topic and published date in relation
 * to the current time
 * @example: Science - 4 hours ago
 * @param story: The story the caption refers to
*/
export const DetailCaption = observer(({ story }: { story: Story }) => {
  const topic = story.primaryTopicName;
  const time = story.publishedDate.fromNow();
  const captionText = topic ? `${time} - ${topic}` : time;

  return (
    <StyledCaption testID="styled-caption">{captionText}</StyledCaption>
  );
});
