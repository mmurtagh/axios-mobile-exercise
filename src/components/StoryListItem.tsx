import { observer } from 'mobx-react-lite';
import React from 'react';
import { NavigationProp, useNavigation } from '@react-navigation/native';
import styled from 'styled-components/native';

import { Story } from '../stores/Story';
import { spacing } from '../utils/styling'
import { Card, Image, Paragraph, Caption } from './components';
import { RootStackParamList } from '..';

const Container = styled.View`
  flex: 3;
  padding-left: ${spacing()};
`;

const ThumbnailImage = styled(Image)`
  flex: 2;  
  align-self: center;
`;

const Author = styled(Caption)`
  margin-top: ${spacing('sm')};
`;

const ListItem = styled(Card)`
  flex-direction: row;
`

const Touchable = styled.TouchableOpacity``;

/** 
 * @name: StoryListItem
 * @description: A list item representing a story
 * @param story: The story the list item represents
 * @param index: Where the story falls in the list
 * @param totalStories: The total number of stories in the list
*/
export const StoryListItem = observer((
  { story, index, totalStories }:
  { story: Story, index: number, totalStories: number}
) => {
  const imageSources: Crop[] = story.getImageSources('4x3');
  const navigation: NavigationProp<RootStackParamList> = useNavigation();

  const accessibilityLabel = `${index + 1} of ${totalStories}: ${story.headline} by ${story.author}`

  return (
    <Touchable
      accessibilityLabel={accessibilityLabel}
      onPress={() => navigation.navigate('StoryDetail', { id: story.id })}
    >
      <ListItem>
        <ThumbnailImage
          testID="thumbnail-image"
          aspectRatio={4 / 3}
          source={imageSources}
        />
        <Container>
          <Paragraph testID="headline">{story.headline}</Paragraph>
          <Author testID="author">{story.author}</Author>
        </Container>
      </ListItem>
    </Touchable>
  );
});
