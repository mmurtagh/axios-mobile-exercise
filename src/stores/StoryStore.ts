import { makeAutoObservable, runInAction } from 'mobx';

import { AxiosService } from '../services/AxiosService';
import { Story } from './Story';

/** 
 * @class: StoryStore
 * @description: A mobx domain store representing a store of Axios stories.
*/
export class StoryStore {
  service: AxiosService;
  isLoadingStories = false;
  stories: Story[] = [];

  constructor() {
    makeAutoObservable(this);
    this.service = new AxiosService();
  }

  /** 
   * @class: loadMostRecentStories
   * @description: Loads the most recent stories using the Axios stream and content APIs.
  */
  async loadMostRecentStories(): Promise<void> {
    const stream: AxiosStream = await this.service.getStream();

    const contentCalls = stream.results.map(id => {
      return this.service.getContent(id);
    });

    const data: AxiosContentInstance[] = await Promise.all(contentCalls);

    runInAction(() => {
      this.stories = data.map((story) => new Story(story))
    });
  }

  getStory(id: string): Story | null {
    return (
      this.stories.find(story => {
        return id === story.id;
      }) || null
    );
  }
}
