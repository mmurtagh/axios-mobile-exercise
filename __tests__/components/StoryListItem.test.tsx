import React from 'react';
import { render } from '@testing-library/react-native';
import { clearTimers } from 'mobx-react-lite';

import { StoryListItem } from '../../src/components/StoryListItem'
import { Story } from '../../src/stores/Story'
import contentInstance from '../data/contentInstance';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe('StoryListItem', () => {
  afterEach(() => {
    clearTimers();
  })

  test('snapshot', () => {
    const story = new Story(contentInstance);
    const obj = render(<StoryListItem story={story} totalStories={20} index={0}/>)

    expect(obj).toMatchSnapshot();
  })

  test('headline', () => {
    const story = new Story(contentInstance);
    const { getByTestId } = render(<StoryListItem story={story} totalStories={20} index={0}/>)

    expect(getByTestId('headline').children[0]).toBe(story.headline);
  })

  test('author', () => {
    const story = new Story(contentInstance);
    const { getByTestId } = render(<StoryListItem story={story} totalStories={20} index={0}/>)

    expect(getByTestId('author').children[0]).toBe(story.author);
  })

  test('image - 4x3 aspect ratio', () => {
    const story = new Story(contentInstance);
    const { getByTestId } = render(<StoryListItem story={story} totalStories={20} index={0}/>)
    const sources = story.getImageSources('4x3');

    expect(getByTestId('thumbnail-image').props.source).toStrictEqual(sources)
  })
});
