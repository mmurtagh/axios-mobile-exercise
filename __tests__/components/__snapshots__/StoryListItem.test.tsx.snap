// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryListItem snapshot 1`] = `
<View
  accessibilityLabel="1 of 20: This is the headline by Foo Barrington"
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 5,
          "flexDirection": "row",
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
          "shadowColor": "rgba(0, 0, 0, 0.25)",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
        },
      ]
    }
  >
    <Image
      aspectRatio={1.3333333333333333}
      source={
        Array [
          Object {
            "height": 100,
            "url": "Image100-4x3",
            "width": 100,
          },
          Object {
            "height": 200,
            "url": "Image200-4x3",
            "width": 200,
          },
          Object {
            "height": 300,
            "url": "Image300-4x3",
            "width": 300,
          },
          Object {
            "height": 400,
            "url": "Image400-4x3",
            "width": 400,
          },
          Object {
            "height": 500,
            "url": "Image500-4x3",
            "width": 500,
          },
        ]
      }
      style={
        Array [
          Object {
            "alignSelf": "center",
            "aspectRatio": 1.3333333333333333,
            "borderRadius": 5,
            "flexBasis": 0,
            "flexGrow": 2,
            "flexShrink": 1,
          },
        ]
      }
      testID="thumbnail-image"
    />
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 3,
            "flexShrink": 1,
            "paddingLeft": 10,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#444",
              "fontFamily": "Helvetica",
              "fontSize": 16,
            },
          ]
        }
        testID="headline"
      >
        This is the headline
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#444",
              "fontFamily": "Helvetica",
              "fontSize": 12,
              "marginTop": 5,
            },
          ]
        }
        testID="author"
      >
        Foo Barrington
      </Text>
    </View>
  </View>
</View>
`;
